go/libraries/doltcore/sqle/diff_table.go

```diff
@@ -341,178 +388,193 @@ func (dp diffPartition) Key() []byte {
+		return nil, errors.New(fmt.Sprintf("table: %s does not exist", tblName))
 	}
 
-	sch, err := t.GetSchema(ctx)
+	wrTblHash, _, err := wr.GetTableHash(ctx, exactName)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	ss, err := schema.NewSuperSchema(sch)
+	cmHash, _, err := cmItr.Next(ctx)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	schRef, err := t.GetSchemaRef()
+	cmHashToTblInfo := make(map[hash.Hash]tblInfoAtCommit)
+	cmHashToTblInfo[cmHash] = tblInfoAtCommit{"WORKING", nil, t, wrTblHash}
+
+	err = cmItr.Reset(ctx)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	h := schRef.TargetHash()
-	addedSchemas := map[hash.Hash]bool{h: true}
+	return &diffPartitions{
+		ctx:             ctx,
+		tblName:         tblName,
+		cmItr:           cmItr,
+		cmHashToTblInfo: cmHashToTblInfo,
+		filterFunc:      filterFunc,
+	}, nil
+}
 
-	diffPartitions := newDiffPartitions(exactName)
-	wrTblHash, _, err := wr.GetTableHash(ctx, exactName)
+// called in a commit iteration loop. Adds partitions when it finds a commit and it's parent that have different values
+// for the hash of the table being looked at.
+func (dp *diffPartitions) processCommit(ctx *sql.Context, cmHash hash.Hash, cm *doltdb.Commit, root *doltdb.RootValue, tbl *doltdb.Table) (*diffPartition, error) {
+	tblHash, _, err := root.GetTableHash(ctx, dp.tblName)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	cmHash, _, err := cmItr.Next(ctx)
+	toInfoForCommit := dp.cmHashToTblInfo[cmHash]
+	cmHashStr := cmHash.String()
+	meta, err := cm.GetCommitMeta()
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	diffPartitions.cmHashToTblInfo[cmHash] = tblInfoAtCommit{"WORKING", nil, t, wrTblHash}
-	err = cmItr.Reset(ctx)
+	ts := types.Timestamp(meta.Time())
+
+	var nextPartition *diffPartition
+	if tblHash != toInfoForCommit.tblHash {
+		partition := diffPartition{toInfoForCommit.tbl, tbl, toInfoForCommit.name, cmHashStr, toInfoForCommit.date, &ts}
+		filterOut, err := dp.filterFunc(ctx, partition)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if !filterOut {
+			nextPartition = &partition
+		}
+	}
+
+	newInfo := tblInfoAtCommit{cmHashStr, &ts, tbl, tblHash}
+	parentHashes, err := cm.ParentHashes(ctx)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, err
+	}
+
+	for _, h := range parentHashes {
+		dp.cmHashToTblInfo[h] = newInfo
 	}
 
+	return nextPartition, nil
+}
+
+func (dp *diffPartitions) Next() (sql.Partition, error) {
 	for {
-		cmHash, cm, err := cmItr.Next(ctx)
+		cmHash, cm, err := dp.cmItr.Next(dp.ctx)
 
 		if err != nil {
-			if err == io.EOF {
-				return ss, diffPartitions, nil
-			}
-
-			return nil, nil, err
+			return nil, err
 		}
 
 		root, err := cm.GetRootValue()
 
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
-		tbl, ok, err := root.GetTable(ctx, exactName)
+		tbl, _, _, err := root.GetTableInsensitive(dp.ctx, dp.tblName)
```

Comment:
What happens if the table doesn't exist in this root? Do you have a test for this case?

Also, this doesn't track tables across renames, which would be nice to do. Might add a TODO.

Reply:
If a table doesn't exist it is equal to nil.  This is hit in every test case you write because the table is never there for the entire history of the db.